Resolution bucket

ruseinov opened the pull request:
Fixes #669

ethanfrey commented on x/gov/bucket.go:
```diff
@@ -59,7 +59,7 @@ const indexNameAuthor = "author"
 func NewProposalBucket() *ProposalBucket {
 	b := migration.NewBucket(packageName, "proposal", orm.NewSimpleObj(nil, &Proposal{})).
-		WithIndex(indexNameAuthor, indexAuthor, false)
+		WithIndex(indexNameElectorate, indexAuthor, false)
```
No need to change this index as part of this PR

ethanfrey commented on x/gov/bucket.go:
```diff
@@ -101,6 +101,51 @@ func (b *ProposalBucket) Update(db weave.KVStore, id []byte, obj *Proposal) erro
+
+func NewResolutionBucket() *ResolutionBucket {
+	b := migration.NewBucket(packageName, "resolution", orm.NewSimpleObj(nil, &Resolution{})).
+		WithIndex(indexNameAuthor, indexElectorate, false)
```
I think you wanted to change it here... indexNameElectorate

ruseinov replied:
Yeah, const definitions that are not grouped on top of the file confuse me, so I think I just changed the wrong reference. The good thing is it will all pop up in tests. Thanks for catching that!

ethanfrey commented on x/gov/bucket.go:
```diff
@@ -101,6 +101,51 @@ func (b *ProposalBucket) Update(db weave.KVStore, id []byte, obj *Proposal) erro
+}
+
+func NewResolutionBucket() *ResolutionBucket {
+	b := migration.NewBucket(packageName, "resolution", orm.NewSimpleObj(nil, &Resolution{})).
```
Hmmm... could we use the new ModelBucket here?

ruseinov replied:
I was thinking about that actually, this is just a sample draft of my changes.

ethanfrey commented on x/gov/codec.proto:
```diff
@@ -133,6 +133,13 @@ message Proposal {
+// Resolution contains TextResolution and an electorate reference.
+message Resolution {
+  weave.Metadata metadata = 1;
+  orm.VersionedIDRef electorate_ref = 2 [(gogoproto.nullable) = false];
```
I think I also requested the original proposal_id to be included here

ruseinov replied:
I'm using the original ticket description as reference as it was not clear what takes precedence. 
If it does not reflect the requirements - may I ask you to update it with latest?

ruseinov replied:
Yep, true. This is still early stages WIP and I committed what I had to fix another PR.

ethanfrey commented on x/gov/codec.proto:
```diff
@@ -224,7 +231,8 @@ message TallyMsg {
 message TextResolutionMsg {
   weave.Metadata metadata = 1;
-  string resolution = 2;
+  bytes electorate_id = 2 [(gogoproto.customname) = "ElectorateID"];
```
This message cannot contain this ref.
The ref is stored in the Proposal, and is the Electorate that actually voted on it.
This one added here is arbitrary.

Hey devs, let's hold an election, stating that the board of directors said "all devs get a pay raise".

Clearly not good. Now, if it recorded that all devs stated "all devs get a pay raise", fair enough, as they said it.

ruseinov replied:
So what you are saying is that we are to use proposal to fetch the electorate ref for Resolution? 
Fair

ethanfrey commented on spec/gogo/x/gov/codec.proto:
```diff
@@ -224,7 +232,8 @@ message TallyMsg {
 message TextResolutionMsg {
   weave.Metadata metadata = 1;
-  string resolution = 2;
+  bytes proposal_id = 2 [(gogoproto.customname) = "ProposalID"];
```
The proposal id is also auto-assigned. The point is we need to pass in some data post-election-tally to the executor so it knows why it is being executed.

I would add something to the context we pass to the Executor. And add a public accessor to x/gov to read that info (so any handler can choose to be gov-aware). Passing proposal_id (or the Proposal object itself) should be enough, as they can query the rest of the context from that object.

ethanfrey commented:
Maybe I can push a commit or two to show a draft of addressing this.
The info *cannot* come from an *untrusted* message. It must come from the contract itself

ethanfrey commented:
I pushed some commits to this, it is basically done.

Main issue is that ProposalID is not stored in the Resolution (not passed in the context). I was thinking how to best pass that along with the Proposal object. (change contents of context).

I think some unit tests just on the TextResolutionHandler (rather than through it's interaction with TallyHandler) would be helpful

ruseinov commented on x/gov/bucket.go:
```diff
@@ -101,6 +101,51 @@ func (b *ProposalBucket) Update(db weave.KVStore, id []byte, obj *Proposal) erro
+	}
+}
+
+func indexElectorate(obj orm.Object) ([]byte, error) {
```
it is actually electorate index, similar to indexProposal. I named it like this for consistency. It is not a refID
index, because we only index the ID and not the version. So I think it is a valid name.

husio replied:
I agree that this is returning a value used for an electorate index building. I think this function does not build the index and in my opinion, the current name does not represent what the function does.

It is a common practice to name indexes after the fields that are used to build it. In case of a single field index, I would expect a name like `field_name_idx`. That is why I think a name for those variables could be improved.

ethanfrey replied:
I saw the name `chainAddressIndexer` in other parts of the code for such a function.
How about renaming to `electorateIDIndexer`?

ruseinov commented on x/gov/handler.go:
```diff
@@ -595,22 +598,36 @@ func (h TextResolutionHandler) Check(ctx weave.Context, db weave.KVStore, tx wea
+	}
+	resolution := &Resolution{
+		Metadata: msg.Metadata,
+		// Note: right here is where I wish to have ID in the model
```
@ethanfrey @husio why don't we introduce a proposalID into the context as part of this PR?
I see no problem introducing a `withProposalID` method which is what I will do and then we can discuss alternatives.
If we want to store the ID - we have to do it now instead of creating a ticket for later?

ruseinov replied:
@husio this has been done in the last commit, I have introduced a wrapper for propsal to save and retrieve it with an ID. I think it looks acceptable.

husio replied:
I think using context to pass temporary variables is a misuse of context. I would question existing context getter/setter if I undestand the code better :see_no_evil: 


I am sharing opinion with what is described [in this blog post](https://peter.bourgon.org/blog/2016/07/11/context.html) - temporary variables should not be inserted into the context.

ruseinov replied:
Well, we just added an id there, I'd like to postpone the discussion about the general approach of storing a proposal in the context to some other time as this has already been done before and this PR just enhances that with an ID. But I get where you are coming from.

ethanfrey replied:
@husio I have read this blog post.

And I have worked with a number of go frameworks that did add important info to the context (including stuff like db references). We pass a number of tx-specific variables there, which are considered optional by most of the code, and the essential ones (db, tx) are passed explicitly.

While I agree with you in theory, I have often struggled to find any good generic way to do without them. (And large nesting on anonymous functions only works if they are defined in one place and is h**l for code-reuse in my experience)

ethanfrey replied:
@ruseinov I am fine with adding a second context ProposalID along with Proposal. And we clean it up later.

Let's get this working now, I will prepare some more PRs

ruseinov commented on x/gov/handler.go:
```diff
@@ -595,22 +598,36 @@ func (h TextResolutionHandler) Check(ctx weave.Context, db weave.KVStore, tx wea
+		return nil, errors.Wrap(errors.ErrNotFound, "no proposal set for passed resolution")
+	}
+	resolution := &Resolution{
+		Metadata: msg.Metadata,
```
I think you are right. I'm going to change that

ethanfrey replied:
So, we should explicitly set it to 1?

I had that first, then figured the other approach was better.
Thank you for the feedback. It would be great if you could add some tips on usage in the `migrations` package. I am still at a loss sometimes on how to use it properly

husio replied:
I have no tips, I think we are still figuring this out :see_no_evil: 

I don't have any better idea than hardcoding it to 1. This might require a manual change when upgraded. I think this is better than an automatic upgrade

ethanfrey replied:
I read the migrations/doc.go and the empty Metadata, with schema to be filled by the Bucket approach was recommended.

I think we have too little experience to see if it is the *right* approach, but at least we can have a *consistent* approach, which makes future changes and testing easier

ruseinov commented on x/gov/model.go:
```diff
@@ -299,6 +300,29 @@ func (m *ProposalCommon) Tally() error {
+	if err := r.Metadata.Validate(); err != nil {
+		return errors.Wrap(err, "invalid metadata")
+	}
+	// TODO: validate proposal ID when we set it
```
going to address that now

ruseinov commented:
@husio @ethanfrey I have addressed all the comments that made sense to me, which is everything except the index name. 
Please review and see that the new way of getting/setting of a proposal makes sense.

ruseinov commented:
actually, there are some tests missing - going to introduce those first.

ruseinov commented:
Tests are done - this seems to be good to go now.

ethanfrey commented on x/gov/context.go:
```diff
@@ -46,3 +52,13 @@ func (a Authenticate) HasAddress(ctx weave.Context, addr weave.Address) bool {
+// CtxProposal reads the context from the proposal (only set for executor)
+func CtxProposal(ctx weave.Context) (*Proposal, []byte) {
+	val, _ := ctx.Value(contextKeyProposal).(proposalWrapper)
+	return val.proposal, val.proposalID
```
This will panic if not set, the previous version set nil.

Please:
1. check if val == nil and return nil, nil explicitly
2. explain the meaning of the []byte return value in the godoc of the function

ruseinov replied:
It does not panic

ethanfrey replied:
Sorry, I missed the cast... `.(proposalWrapper)`, not a pointer.
You are right, my bad.

ethanfrey commented on x/gov/handler.go:
```diff
@@ -217,7 +217,8 @@ func (h TallyHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx)
 	// ensure that the gov.Authenticator is used in those Handlers
-	voteCtx := withElectionSuccess(ctx, common.ElectionRuleRef.ID)
+	// we also add the proposal with id that was passed that can be accessed via CtxProposal()
+	voteCtx := withProposal(withElectionSuccess(ctx, common.ElectionRuleRef.ID), proposal, msg.ProposalID)
```
:+1:

ethanfrey commented on x/gov/handler_test.go:
```diff
@@ -524,6 +524,71 @@ func TestDeleteProposal(t *testing.T) {
+			}
+
+			if spec.WantDeliverErr != nil {
+				return // skip further checks on expected error
```
Maybe check that the resolution was *not created* ?

ethanfrey commented on x/gov/model.go:
```diff
@@ -299,6 +300,31 @@ func (m *ProposalCommon) Tally() error {
+	if err := r.Metadata.Validate(); err != nil {
+		return errors.Wrap(err, "invalid metadata")
+	}
+	if len(r.ProposalID) == 0 {
```
We can enforce 8 bytes as only valid number from IDGenBucket
